### DISCUSSION TITLE
Writing a plugin that supports dark mode?

### DISCUSSION THREAD

**Discussion by emilmuller** [tailwindlabs/tailwindcss#11945] [Help] [answered]
Hi,

I couldn't find it in the documentation. How can I write a plugin that supports dark mode? E.g. I want to add a component .mycomponent that looks different depending on which mode is active.

I thought I would simply target prefers-color-scheme CSS media query as well as .dark class, but then again, I would really like to support whatever is chosen in 'darkMode' in the config file.

Kind regards, Emil

**Comment by wongjn** [comment, marked as answer]
I believe the Tailwind way would to have the two stylings as different classes, and let the consumer implement the dark mode by prefixing one of the classes with `dark:`. This is because it would be difficult for a consumer to "undo" the dark mode styles if they did not want them to change in dark mode. Having the consumer explicitly specify the dark mode also gives a better idea of what is going on. For example, `@tailwindcss/typography` (an offical Tailwind plugin), follows this concept by [having a `prose-invert` class that you would couple with the `dark:` variant](https://tailwindcss.com/docs/typography-plugin#adapting-to-dark-mode).

Otherwise, to answer your request directly, consider emulating the [core Tailwind `darkMode` plugin implementation](https://github.com/tailwindlabs/tailwindcss/blob/696810fd5a3d30ee8225913af268be1d2c9dbced/src/corePlugins.js#L216-L231) in your own plugin like:

```js
require('tailwindcss/plugin')(({ addComponents, config }) => {
  let [mode, className = '.dark'] = [].concat(config('darkMode', 'media'));
  if (mode === false) {
    mode = 'media';
  }

  const darkContext = mode === 'media'
    ? '@media (prefers-color-scheme: dark)'
    : `:is(${className} &)`;

  addComponents({
    '.my-component': {
      backgroundColor: 'red',
      [darkContext]: {
        backgroundColor: 'blue'
      },
    },
  });
})
```

See it working in [this Tailwind Play](https://play.tailwindcss.com/8ghuduiFsj?file=config).

**Reply by emilmuller** [reply]
Thanks! Exactly what I needed
